pxtcompiler/emitter/languageservice.ts

```diff
@@ -0,0 +1,650 @@
+            type: r.isMemberCompletion ? "memberCompletion" : "identifierCompletion"
+        };
+
+        let resultSymbols: CompletionSymbol[] = []
+
+        let tsPos: number;
+        if (isPython) {
+            // for Python, we need to transpile into TS and map our location into
+            // TS
+            const res = transpile.pyToTs(opts)
+            if (res.syntaxInfo && res.syntaxInfo.symbols) {
+                resultSymbols = completionSymbols(res.syntaxInfo.symbols, COMPLETION_DEFAULT_WEIGHT);
+            }
+            if (res.globalNames)
+                lastGlobalNames = res.globalNames
+
+            if (!resultSymbols.length && res.globalNames) {
+                resultSymbols = completionSymbols(pxt.U.values(res.globalNames), COMPLETION_DEFAULT_WEIGHT)
+            }
+
+            // update our language host
+            Object.keys(res.outfiles)
+                .forEach(k => {
+                    if (k === tsFilename) {
+                        host.setFile(k, res.outfiles[k])
+                    }
+                })
+
+            // convert our location from python to typescript
+            if (res.sourceMap) {
+                const pySrc = src
+                const tsSrc = res.outfiles[tsFilename] || ""
+                const srcMap = pxtc.BuildSourceMapHelpers(res.sourceMap, tsSrc, pySrc)
+
+                const smallest = srcMap.py.smallestOverlap(span)
+                if (smallest) {
+                    tsPos = smallest.ts.startPos
+                }
+            }
+
+            // filter based on word match if we get too many (ideally we'd leave this filtering for monaco as it's
+            // better at fuzzy matching and fluidly changing but for performance reasons we want to do it here)
+            if (!isMemberCompletion && resultSymbols.length > MAX_SYMBOLS_BEFORE_FILTER) {
+                resultSymbols = resultSymbols
+                    .filter(s => (isPython ? s.symbol.pyQName : s.symbol.qName).indexOf(partialWord) >= 0)
+            }
+        } else {
+            tsPos = position
+            opts.ast = true;
+            host.setOpts(opts)
+            const res = runConversionsAndCompileUsingService()
+        }
+
+        const prog = service.getProgram()
+        const tsAst = prog.getSourceFile(tsFilename)
+        const tc = prog.getTypeChecker()
+        let tsNode = findInnerMostNodeAtPosition(tsAst, tsPos);
+        const commentMap = pxtc.decompiler.buildCommentMap(tsAst);
+
+        // abort if we're in a comment
+        const inComment = commentMap.some(range => range.start <= position && position <= range.end)
+        if (inComment) {
+            return r;
+        }
+
+        // abort if we're in a string literal
+        const stringLiteralKinds = [SK.StringLiteral, SK.FirstTemplateToken, SK.NoSubstitutionTemplateLiteral];
+        const inLiteral = stringLiteralKinds.some(k => tsNode.kind === k)
+        if (inLiteral) {
+            return r;
+        }
```

Comment:
`const commentMap [...]` until here is new. We abort if we're in a multi-line comment or string literal

Reply:
Is there a perf hit here? I suppose the type checking / etc that will be running will take significantly longer than a linear pass over the file so it should be fine.

Reply:
Yeah, it really should be incremental. As should most of the language service.

I'll measure the perf hit.
- [ ] TODO

Reply:
<10ms for PlanetPuttPutt, so barely measurable